### ISSUE TITLE
[speech-to-text] web sockets, bufferedAmount

### ISSUE THREAD

**Issue by deadplant** [watson-developer-cloud/node-sdk#285]
It looks like recognize_stream.js tries to use socket.bufferedAmount for flow control.  However, the included W3CWebSocket.js does not implement it.  The value is always zero according the comments in W3CWebSocket.js which matches what I'm seeing in practice.

`"name": "watson-developer-cloud",  "version": "1.9.4",`

`watson-developer-cloud/services/speech_to_text/recognize_stream.js:`
```js
// flow control - don't ask for more data until we've finished what we have
// todo: see if this can be improved
RecognizeStream.prototype.afterSend = function afterSend(next) {
  if (this.socket.bufferedAmount <= this._writableState.highWaterMark || 0) {
    next();
  } else {
    setTimeout(this.afterSend.bind(this, next), 10);
  }
};
```

watson-developer-cloud/node_modules/websocket/lib/W3CWebSocket.js line 46:
```this._bufferedAmount = 0;  // Hack, always 0.```

**Comment by nfriedly** [comment]
Oh, interesting. So it's not controlling the flow at all, and could lead to a lot of memory use in the case of a large file and a slow upload speed :(

It's also checking `bufferedAmount` vs `_bufferedAmount`, I wonder if that makes a difference.

Do you have any thoughts on how to improve it? (Well, I suppose we could send a PR to websocket.)

I'm using the W3C version because I intend to have this work in browsers. (Right now there's a [separate library](https://github.com/watson-developer-cloud/speech-javascript-sdk), but I'd like to merge the two.)

**Comment by deadplant** [comment]
Ah well... I don't really feel qualified to propose a solution.  I'm just a hack! :)

In an attempt to get my thing working right now I am looking at my_recognize_stream_instance.socket._connection.outputBufferFull and if it is 'true', pausing my writes.  Then, resuming my writes on the 'drain' event emitted by socket._connection.

This should cause the web sockets's buffer to fill, then flush, then fill, then flush... hardly optimal but enough for me to muddle along for the moment.
I haven't located an accurate bufferedAmount value somewhere in the web socket library yet but I'll keep looking.

**Comment by nfriedly** [comment]
Sounds good. 

Just out of curiosity, why did you start digging into this?

**Comment by deadplant** [comment]
I am hoping to make use of the speech-to-text service to reduce some labour here.

I need to run through chunks of audio that are larger/longer than the watson session upload limit.   So, rather than using the .pipe() chain from the example, I am reading from my source (ffmpeg output) into a buffer, and asynchronously writing to the RecognizeStream._write() function.
When I approach the watson data limit I stop writing to it, wait for it to close, fire up a fresh one and resume writing to the new recognizeStream.

So... of course during this downtime I am building up a backlog of realtime input audio.  I then slammed all that data in to the new recognizeStream, triggering the issue.
Also, because I am using the 'callback' parameter of RecognizeStream._write(chunk, encoding, callback)  to trigger my next attempt to write to watson I was actually getting a "call stack size limit exceeded" crash whenever I wrote to it too fast.

**Comment by nfriedly** [comment]
oh, yikes.

I might be able to make the callback async, which should avoid the call stack size problems.

But, outside of that, I think you can also have multiple streams going at once. So when the first one is approaching the limit, let it close gracefully, but start sending your audio to the next stream immediately.

Or you could keep doing what you're doing now, since that works well enough for the moment. 

Oh, and BTW, the limit is in megabytes, so if you encode things with FLAC (lossless), then you can get a lot more audio sent before hitting the limit (vs WAV, anyways - ogg/opus is also supported, it compresses better than FLAC but it's lossy.) That might be the best answer to your issue.

**Comment by deadplant** [comment]
Yeah.  I was hoping I could keep it to one session at a time just to keep the code simple.  The gaps aren't a big deal unless watson is strictly limited to real-time processing... in which case I will fall slowly further behind each time.   I was hoping it could catch up a bit when I send it data slightly faster than realtime at the start of each new session.   I haven't measured this yet.

If watson is strictly realtime I will have to fire up session B as I approach the limit for session A so it is ready the moment I stop writing to A.  Then let A close gracefully while already writing to B.   The issue being A's final speech-to-text sentence result can take some time to show up after I finish writing.
so a gap-less hand-over of writes would mean overlapping 'results' coming in which i would have to keep in order.

I would love to send FLAC, however... my source is an effectively continuous ffmpeg output.   With pcm_s16le I can break up the stream of data and write valid audio data to successive watsons easily whereas with FLAC I probably can't just split the audio data stream anywhere I want.
I can think of a few ways to handle splitting up the flac stream but since it would only mitigate, not eliminate the need for rolling-watson-sessions I'm leaving it for "v2"
(Also watson's word-timestamps start at 0 for each fresh session, so I have to know precisely how many seconds of audio I've sent in previous sessions so I can add that offset to new results...  using raw audio lets me to that calculation with nothing more than the number of bytes sent which is handy)

**Comment by nfriedly** [comment]
Ok, fair points.

It's not strictly real-time, it'll process the audio as fast as it comes in, so the catching up plan should work. I'll look into improving the other things and probably file an issue on websocket asking about the feasibility of an accurate `bufferedAmount`

**Comment by nfriedly** [comment]
BTW, I just changed the `next()` call to `process.nextTick(next)` - I think that will fix the call stack exceeded issues. It should be released in v1.12.3 once Travis CI finishes. Is it easy for you to roll back to where it was crashing and test this out for me? (Even if your current solution proves to be better, I still would like to know if I'm avoiding the crashes now.)

**Comment by nfriedly** [comment]
I'm going to close this for now; please re-open or file a new ticket if you're still experiencing issues.
